tests/Functional/Driver/Mysqli/ResultTest.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Tests\Functional\Driver\Mysqli;
+
+use Doctrine\DBAL\Driver\Mysqli\Result;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+use Doctrine\DBAL\Tests\TestUtil;
+use mysqli;
+use mysqli_driver;
+use mysqli_sql_exception;
+
+use function assert;
+use function sprintf;
+
+use const MYSQLI_REPORT_ERROR;
+use const MYSQLI_REPORT_OFF;
+use const MYSQLI_REPORT_STRICT;
+
+/** @requires extension mysqli */
+final class ResultTest extends FunctionalTestCase
+{
+    private const TABLE_NAME = 'tesult_test_table';
+
+    private mysqli $nativeConnection;
+
+    protected function setUp(): void
+    {
+        if (! TestUtil::isDriverOneOf('mysqli')) {
+            self::markTestSkipped('This test requires the mysqli driver.');
+
+            return;
+        }
+
+        $nativeConnection = $this->connection->getNativeConnection();
+
+        assert($nativeConnection instanceof mysqli);
+
+        $this->nativeConnection = $nativeConnection;
+
+        $table = new Table(self::TABLE_NAME);
+        $table->addColumn('my_col_1', 'integer', ['notnull' => true]);
+
+        $this->dropAndCreateTable($table);
+    }
+
+    protected function tearDown(): void
+    {
+        $this->dropTableIfExists(self::TABLE_NAME);
+    }
+
+    public function testSuccessfulRowCountFromAffectedRows(): void
+    {
+        $mysqliStmt = $this->nativeConnection
+            ->prepare(sprintf('INSERT INTO %s VALUES (7);', self::TABLE_NAME));
+        $mysqliStmt->execute();
+
+        self::assertSame(1, $mysqliStmt->affected_rows);
+        self::assertSame(1, (new Result($mysqliStmt))->rowCount());
+    }
+
+    public function testFailingRowCountFromAffectedRows(): void
+    {
+        $mysqliStmt = $this->nativeConnection
+            ->prepare(sprintf('INSERT INTO %s VALUES (NULL);', self::TABLE_NAME));
+
+        $mysqliDriver = new mysqli_driver();
+
+        // Ensure default configuration present in PHP < 8.1.
+        $mysqliDriver->report_mode = MYSQLI_REPORT_OFF;
+
+        $mysqliStmt->execute();
+
+        self::assertSame(-1, $mysqliStmt->affected_rows);
+        self::assertSame(-1, (new Result($mysqliStmt))->rowCount());
+
+        // Ensure default configuration present in PHP >= 8.1.
+        // @see: https://php.watch/versions/8.1/mysqli-error-mode.
+        $mysqliDriver->report_mode = MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT;
+
+        $error = null;
+
+        try {
+            $mysqliStmt->execute();
+        } catch (mysqli_sql_exception $error) {
+            // no-op.
+        }
```

Comment:
Because this is a test, we can let either let an exception happen or let the test explicitly fail in the catch with a reason what went wrong if it's relevant for the test case itself. I assume the latter is the case you seek because you check for the exception there.

Reply:
I think this is not possible in this case, because I don't want the test to fail at this point, but later on the creation of the `Result` instance.

Reply:
Is it even supposed to throw a `mysqli_sql_exception` in the constructor? It does, sure, but I'd expect that a Doctrine Exception is thrown in there.

Reply:
I agree. In part, that's why I'm proposing to avoid a `-1` response to represent an error, but throw a domain exception instead.
This PR is not changing any behavior, but trying to provide a clear understanding and confirmation about the current flow.